File: Java/binary-tree-inorder-traversal.java
/**
 * Problem Link: https://leetcode.com/problems/binary-tree-inorder-traversal/
 * 
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        HashSet<TreeNode> visited = new HashSet<TreeNode>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        List<Integer> list = new ArrayList<Integer>();
        if(root == null) return list;
        stack.push(root);
        while(!stack.empty()) {
            TreeNode node = stack.peek();
            if(node.left != null && !visited.contains(node.left)) {
                stack.push(node.left);
            } else {
                node = stack.pop();
                list.add(node.val);
                visited.add(node);
                if(node.right != null) stack.push(node.right);
            }
        }
        return list;
    }
    
    public List<Integer> inorderTraversal(TreeNode root) {
        HashSet<TreeNode> visited = new HashSet<TreeNode>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        List<Integer> list = new ArrayList<Integer>();
        if(root == null) return list;
        stack.push(root);
        while(!stack.empty()) {
            TreeNode node = stack.peek();
            if(node.left != null) {
                if(visited.contains(node.left)) {
                    node = stack.pop();
                    list.add(node.val);
                    visited.add(node);
                    if(node.right != null) stack.push(node.right);
                } else {
                    if(node.left != null) {
                        stack.push(node.left);
                    } else {
                        node = stack.pop();
                        list.add(node.val);
                        visited.add(node);
                        if(node.right != null) stack.push(node.right);
                    }
                }
            } else {
                node = stack.pop();
                list.add(node.val);
                visited.add(node);
                if(node.right != null) stack.push(node.right);
            }
        }
        return list;
    }
}

File: Swift/binary-tree-inorder-traversal.swift
/**
 * https://leetcode.com/problems/binary-tree-inorder-traversal/
 * 
 * 
 */ 
// Date: Sun May 10 17:10:54 PDT 2020
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    /// Recursive solution.
    func inorderTraversal(_ root: TreeNode?) -> [Int] {
        guard let root = root else { return [] }
        return self.inorderTraversal(root.left) + [root.val] + self.inorderTraversal(root.right)
    }
}
